lib/ansible/parsing/dataloader.py

```diff
@@ -233,35 +236,33 @@ def path_dwim(self, given):
         given = unquote(given)
         given = to_text(given, errors='surrogate_or_strict')
 
-        if given.startswith(u"/"):
-            return os.path.abspath(given)
-        elif given.startswith(u"~"):
-            return os.path.abspath(os.path.expanduser(given))
+        if given.matches(os.path.sep) or given.matches(u'~'):
+            path = unfrackpath(given, follow=False)
         else:
             basedir = to_text(self._basedir, errors='surrogate_or_strict')
-            return os.path.abspath(os.path.join(basedir, given))
+            path = unfrackpath(os.path.join(basedir, given), follow=False)
+
+        return path
 
     def _is_role(self, path):
-        ''' imperfect role detection, roles are still valid w/o main.yml/yaml/etc '''
+        ''' imperfect role detection, roles are still valid w/o tasks|meta/main.yml|yaml|etc '''
 
-        isit = False
         b_path = to_bytes(path, errors='surrogate_or_strict')
-        b_upath = to_bytes(unfrackpath(path), errors='surrogate_or_strict')
-
-        for suffix in (b'.yml', b'.yaml', b''):
-            b_main = b'main%s' % (suffix)
-            b_tasked = b'tasks/%s' % (b_main)
-
-            if (
-                b_path.endswith(b'tasks') and
-                os.path.exists(os.path.join(b_path, b_main)) or
-                os.path.exists(os.path.join(b_upath, b_tasked)) or
-                os.path.exists(os.path.join(os.path.dirname(b_path), b_tasked))
-            ):
-                isit = True
-                break
-
-        return isit
+        b_upath = to_bytes(unfrackpath(path, follow=False), errors='surrogate_or_strict')
+
+        for finddir in (b'meta', b'tasks'):
+            for suffix in (b'.yml', b'.yaml', b''):
+                b_main = b'main%s' % (suffix)
+                b_tasked = b'%s/%s' % (finddir, b_main)
+
+                if (
+                    RE_TASKS.search(path) and
```

Comment:
Since the new code uses RE_TASKS in other places, probably put the comment about what the intended matches should be at the top where RE_TASKS is defined.